x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -234,16 +188,82 @@ public void finalizeSnapshot(
+        }
+    }
+
+    @Override
+    public void endVerification(String packedSeed) {
+        // verification bypasses the encrypted blobstore because it works easier without a password generation
+        String seed = packedSeed.substring(0, packedSeed.indexOf('.'));
+        super.endVerification(seed);
+    }
+
+    @Override
+    public void verify(String packedSeed, DiscoveryNode localNode) {
+        // verification bypasses the encrypted blobstore because it works easier without a password generation
+        String seed = packedSeed.substring(0, packedSeed.indexOf('.'));
+        String uuid = packedSeed.substring(seed.length() + 1, packedSeed.lastIndexOf('.'));
+        String hash = packedSeed.substring(packedSeed.lastIndexOf('.') + 1);
+        try {
+            String computedHash = AESKeyUtils.computeId(AESKeyUtils.generatePasswordBasedKey(repositoryPassword, uuid));
+            if (false == hash.equals(computedHash)) {
+                throw new ElasticsearchException("Password mismatch on node {}", localNode);
+            }
+        } catch (Exception e) {
+            throw new RepositoryVerificationException(metadata.name(), "Error verifying password hash", e);
+        }
+        super.verify(seed, localNode);
+    }
+
+    @Override
+    public BlobContainer testBlobContainer(String seed) {
+        // bypass encryption for test blobs because the lifecycle of password generation is a bit complex
+        // when accounting for optional repository verification and it is not very useful
+        EncryptedBlobStore encryptedBlobStore = (EncryptedBlobStore) blobStore();
+        return encryptedBlobStore.delegatedBlobStore.blobContainer(basePath().add(testBlobPrefix(seed)));
+    }
+
+    @Override
+    public void snapshotShard(SnapshotShardContext context) {
+        if (false == licenseStateSupplier.get().isAllowed(XPackLicenseState.Feature.ENCRYPTED_SNAPSHOT)) {
+            context.onFailure(LicenseUtils.newComplianceException("encrypted snapshots"));
             return;
         }
         super.snapshotShard(context);
     }
 
+    @Override
+    public void executeConsistentStateUpdate(
+        BiConsumer<RepositoryData, ActionListener<ClusterStateUpdateTask>> createUpdateTaskAsync,
+        String source,
+        Consumer<Exception> onFailure
+    ) {
+        super.executeConsistentStateUpdate((repositoryData, createUpdateTaskListener) -> {
+            if (repositoryData.equals(RepositoryData.EMPTY) && false == isReadOnly()) {
+                threadPool().generic().execute(ActionRunnable.wrap(createUpdateTaskListener, (l) -> {
+                    // TODO this is not safe for master fail-over
+                    // TODO this needs the publish, write, publish pattern to ensure it doesn't get lost
+                    ((EncryptedBlobStore) blobStore()).maybeInitializePasswordGeneration();
```

Comment:
This must eventually use the same mechanism as the password change. Both generate a new password generation.
* publish the intention in the cluster state
* write the generation marker blob (described by the intention in the cluster state)
* remove the intention from the cluster state

The repository cannot be removed while an intention is published.